server/src/instant/grouped_queue.clj

```diff
@@ -1,334 +1,162 @@
+
+     :max-workers  :: long | nil
+
+   If exectutor is not provided, ~ cached thread pool will be created with at most this many threads.
+
+     :metrics-path :: String | nil
+
+   A string to report gauge metrics to. If skipped, no reporting"
+  [{:keys [group-key-fn combine-fn process-fn executor max-workers metrics-path]}]
+  (let [groups      (ConcurrentHashMap.)
+        accepting?  (atom true)
+        processing? (atom true)
+        items       (AtomicInteger. 0)
+        workers     (AtomicInteger. 0)
+        executor    (cond
+                      (some? executor)
+                      executor
+
+                      config/fewer-vfutures?
+                      (ThreadPoolExecutor. 0 (or max-workers 2) 1 TimeUnit/SECONDS (LinkedBlockingQueue.))
+
+                      :else
+                      (Executors/newVirtualThreadPerTaskExecutor))
+        cleanup-fn  (when metrics-path
+                      (gauges/add-gauge-metrics-fn
+                       (fn [_]
+                         [{:path  (str metrics-path ".size")
+                           :value (AtomicInteger/.get items)}
+                          (when-some [t (longest-wait-time groups)]
+                            {:path  (str metrics-path ".longest-waiting-ms")
+                             :value t})
+                          {:path (str metrics-path ".worker-count")
+                           :value (AtomicInteger/.get workers)}])))
+        shutdown-fn (fn [{:keys [timeout-ms]
+                          :or {timeout-ms 1000}}]
+                      (when cleanup-fn
+                        (cleanup-fn))
+                      (reset! accepting? false)
+                      (ExecutorService/.shutdown executor)
+                      (if (ExecutorService/.awaitTermination executor timeout-ms TimeUnit/MILLISECONDS)
+                        :shutdown
+                        (do
+                          (reset! processing? false)
+                          (if (ExecutorService/.awaitTermination executor timeout-ms TimeUnit/MILLISECONDS)
+                            :shutdown
+                            (do
+                              (ExecutorService/.shutdownNow executor)
+                              :terminated)))))]
+    {:group-key-fn (or group-key-fn identity)
+     :combine-fn   (or combine-fn (fn [_ _] nil))
+     :process-fn   process-fn
+     :groups       groups
+     :accepting?   accepting?
+     :processing?  processing?
+     :items        items
+     :workers      workers
+     :executor     executor
+     :shutdown-fn  shutdown-fn}))
+
+(defn stop
+  "Stops grouped queue. Shuts executor down. Possible options:
+
+     :timeout-ms :: long
+
+   How long to wait for existing tasks to finish processing before interrupting."
+  ([q]
+   ((:shutdown-fn q) {}))
+  ([q opts]
+   ((:shutdown-fn q) opts)))
+
+(defn size
+  "~ Amount of items currently in all queues"
+  [q]
+  (AtomicInteger/.get (:items q)))
```

Comment:
Nit: should we call this `num-items` or `size`? Reasoning: when I read `items` I expect it to be a list of something, but this is an AtomicInteger

Reply:
Good point!